### DISCUSSION TITLE
Official roles package (Roles Work Group)

### DISCUSSION THREAD

**Discussion by StorytellerCZ** [meteor/meteor#13357] [Feature Requests]
## Members

- @StorytellerCZ 
- @jankapunkt 
- @denihs 

**Package to be migrated to core**: `alanning:roles`

**Deadline:** Meteor 3.1 (October 2024)

## Steps
1. Move `alanning:roles` to core packages
2. Integrate tests
3. Documentation, guide and examples
4. Other additions

## Other considerations
- https://packosphere.com/pwix/roles
- Validated method mixin
- jam:method
- integration into accounts package (after user creation, assign a default role as specified by the setting)

**Comment by jkasevits** [comment]
Hi. I'm Vooteles from the forums.

In my opinion alanning:roles is a very nice package but it would not be a good idea to bring it into Meteor's core and it would be much better to leave it (and any other package similar to it) as a separate community maintained package. For the following reasons:

1. Unlike authentication (i.e. the accounts system in Meteor), authorization (who has the right to do what) needs tend to be quite different from app to app. alanning:roles has made many choices on how to structure and manage this data. These choices might be suitable for one app, while not a good fit for another. Meteor should not force developers into these premade decisions unless it comes with massive benefits. In this case it does not come with benefits.
2. It would bring unnecessary complexity into Meteor core. alanning:roles does not simply provide roles, but also makes it possible to put them into hierachies, roles can have parents and subroles, there are "scopes" for roles (global roles vs scoped roles) and what not. And a bunch of functions to controll all this data. It also makes assumptions on how users want the roles data to be published. All this stuff is very specific to each app and should be left to developers outside the core to handle (either as an external package such as alanning:roles or custom app code). Having this complexity in core would not help make Meteor a better development platform.
3. The accounts system in Meteor's core is mature and it provides a very stable foundation for packages such as alanning:roles to interface with. Meteor core provides the authentication part and external packages such as alanning:roles can provide the authorization part. alanning:roles has worked well for years and I don't recall many issues where changes in Meteor core broke something in alanning:roles (except the transition from Meteor 2 to Meteor 3 of course, which affected almost everything). Since for this part of Meteor external packages work so well, why change it? alanning:roles is exactly the kind of package that suits perfectly for the Meteor community packages umbrella.
4. Meteor should try to stay as lean as possible and only bring in to the core code that is absolutely necessary for the project and cannot be reasonably maintaned as an external package. Each new line of code adds complexity, creates more work on maintenance and, importantly, makes it more difficult or at least more taunting for new developers to start contributing to the core. The team developing and maintaining Meteor is small. This should be considered each time there is an idea to add something to the core.
5. Having authorization in external packages keeps the door open for new, competing implementations in the form of other external packages. This is largely what open source is about. When one implementation is in core there is a lot less motivation for different packages to appear. Over time this can have a negative effect.

Meteor should provide authentication and external packages should provide authorization. I do not see any benefit in bringing the authorization part into core as well. Only downsides.

**Reply by StorytellerCZ** [reply]
Thank you for your thoughtful reply. I do acknowledge the worries, so I'll try to alleviate them in some cases, in others I feel we will continue to disagree.
1. Yes, authorization is different app to app, but as you have mentioned yourself `alanning:roles` is very flexible and accommodates many authorization models. It has also been recommended in the guide and there aren't any competing packages (even `pwix:roles` is an extension). I do think that integrating into core is a benefit and fits the Meteor ethos of opinionated framework. In the work group we will be exploring other options as well.
2. Honestly I don't get this argument. The roles package is very flexible, but in its base use it is very simple. The flexibility allows for advance uses (`pwix:roles` event further expands on it with hierarchy). How to deal best with publications will be matter of debate, but in this case I think inclusion in core can be of benefit where we can tap into the existing user publication and with a check if the roles package is included also add to it to publish the user's role (unless opted out), which will reduce the need to have an extra publication for roles as it is now. Accounts package itself is also very complex.
3. Why change what works? Why I do appreciate the trust you place in me @jankapunkt @harryadel @copleykj and others who move around the MCP packages, there is currently 62 projects in there. Comparatively here is more resources (people) than in MCP. Roles is also a very mature package that doesn't require much work after the integration is complete and given how a crucial part of many applications it is it will benefit for having first class support when there are breaking changes that affect it, rather than playing catch-up in my (or other people's) spare time.
4. This is a difference in philosophies that has been discussed since 2015. By that reasoning you could argue that the accounts package should be an external package. But like roles, you can include it if you need it. Does it add more lines of code, yes. Does it create more work, I would say temporarily. Given that roles package is pretty much standalone, stable and battle tested, there is an upfront cost of integrating it into core, but after that I don't think the maintenance burden will increase significantly. If that wasn't the case then I doubt that Meteor Software would have wanted it. Also in terms for beginner contributors, I would say that roles package would be one of the more approachable (have you seen the code in accounts package?). I'm not privy to the all the thinking of Meteor Software on this, but from my perspective it is building up on the accounts packages by expanding the default/core functionality to include what most applications will need. Part of that probably is for marketing to corporations where role base access control (https://www.enterpriseready.io/). Sadly people making decisions on these technologies will at best look at official packages and won't care that there is an external roles package. But if they see it there, they might think: "Oh! We don't have to pay for authorization SaaS, neat!". I also have to add here my obligatory pitch to include Apple OAuth into core for similar reason, people doing the evaluation will only look at the official packages/capabilities, if there is anything external that means extra costs for them. And today that becomes and additional issue with supply chain safety as well (which is all the rave at conferences).
5. Going back to my point in 1, there have not been any competing packages. If anything the suppression you are referring to could be said have started with inclusion of `alanning:roles` in the guide. I look at it that for 95% (or more) of apps `alanning:roles` (potentially with a few extra custom rules) is more than enough. For the remainder only very specific custom solution is an option and that will not live beyond the app, in these cases they just don't include the roles package and are happy.

@harryadel pointed out to me that neither Rails or Laravel have build-in roles package either. Honestly I was a bit surprised when Meteor Software came to me that they want this package in particular in core, but from my perspective it makes sense as mentioned above in point 4. I also think that this could become another great defining feature of Meteor.

**Reply by StorytellerCZ** [reply]
Also when roles becomes part of core other packages can build more confidently on top of it. When Blaze, for example, was more integrated there was a boom of packages like [useraccounts](https://atmospherejs.com/useraccounts), different admins and other that gave you a lot of extra functionality out of the box. That is the benefit of more opinionated system (though good package will allow you to accommodate even the non-standard approaches).

**Comment by jankapunkt** [comment]
I'd like to add a general note that we all should be aware that these integrations will cause some frictions in the community, since there are two fundamentally different views on how Meteor should move forward:

1. become more opinionated by default (but remain flexible; zero-config-first)
2. become event less opinionated and more flexible (for example offshore more solutions to npm packages)

Related forums discussions: 
- https://forums.meteor.com/t/make-meteor-great-again/61892
- https://forums.meteor.com/t/is-anyone-still-preferring-blaze-over-react/61996

The new roadmap indicates a strong direction for option 1 and this is one step forward to it.

**Reply by StorytellerCZ** [reply]
I think we can hold an event on Discord to talk about this issue in general. Thoughts?

**Comment by harryadel** [comment]
Ok, I spoke with Jan in private but chose to continue our discussion in public so others can benefit and chime in. Most of my comments resonate with what Vooteles/@jkasevits brought up, but I'll comment on points both of you discussed while bringing up further points of my own.

Vooteles says it's unnecessary complexity, but Jan rebuts that it's already mature and doesn't need extra work. The work needed in the future isn't dependent on further development; even casual maintenance would take a toll. Nobody anticipated the fibers shenanigans - imagine if it were to happen with packages of the caliber of Roles in the core. That would lengthen the work tremendously. Growth and further development is a far-off dream, and even maintenance is no easy task. If we should take a lesson from the fibers situation, it's that it's in our best interest to keep Meteor core small and lean in order to be able to work around any problems that may arise in the future, not to increase our luggage!

I'd also like to highlight the fact that the core team is heavily understaffed. Lol, if they had enough manpower, they wouldn't need these work groups in the first place! For those who don't know, @nachocodoner works part-time while @leonardoventurini and @denihs work full-time. So 2.5 devs for a project the size of Meteor? You think that's going to cut it?

Another personal trivia of mine: when the core team spoke of bringing packages to the core, I thought to myself [collection2](https://github.com/Meteor-Community-Packages/meteor-collection2) is the best candidate. But then I was browsing the forums the other day and noticed [pmogollons:zod-schema](https://forums.meteor.com/t/new-package-pmogollons-zod-schema/62096). It just dawned on me that no matter how good a package is, over time it might erode and new cool alternatives appear on the scene.

By locking into a certain method of doing things, you end up stifling the creativity of other developers and what they choose to do. I know you might now think, "Then what about 'batteries-included' and 'being opinionated'?" Yes, be opinionated, but only about things that matter and the big, broad decisions. If you make decisions on every minute detail, you end up stifling and dragging down the developer. Meteor makes a big decision about what type of database you use but not how you organize your data.

In conjunction with my last point, it's important when you come up with opinionated decisions that you allow ways for developers to circumvent them in case they need to do something different, regardless of how good your solution is. So here we're expending brain power, thinking about it, and discussing things. Meteor already has plenty of problems of its own, and while the fiber-less upgrade is a good move on its own, it's not problem-free and it caused a ruckus not only for developers but for third-party packages out there that still need maintenance. Adding more to Meteor's plate post-fibers is shooting ourselves in the foot.

Lastly, I'm not completely disregarding the talk of including packages in the core, but meteor-roles isn't the one. I think the best candidates for inclusion into the core are simple packages that solve niche problems. Remember [this.unblock](https://github.com/meteorhacks/unblock) and how it got adapted to the core? I think maybe the next optimal candidate is [mongo-collection-instances](https://github.com/Meteor-Community-Packages/mongo-collection-instances).

One of the points that makes it compelling to include a package into the core is how small it is, because even if we made a bad decision, the cost of future maintenance or even removal would be trivial.

**Reply by StorytellerCZ** [reply]
No matter what package you choose, there is always going to be cohort of people who will disagree (my disagreement in the past was about the source from where to get the packages). It goes back to the vision of what Meteor should be. As stated above I'm with the vision where Meteor is more opinionated than not and includes more stuff that businesses need in the official section to make it more appealing. Which is why I'm advocating for Apple OAuth, I personally don't need it, but I do recognize the business value of just having it on the freaking list of things when trying to sell Meteor.
As for the other packages, there has been talk of including other in as well. I've specifically talked with Meteor Software about collection-hooks (which my conclusion was that the current community implementation is not a good fit due to code maintainability/complexity due to having to hack things for being outside of core).
Also what about under-staffing in MCP? There we can barely put together the equivalent of one full-time developer and often not even a part-time. Also there is more likelihood of Meteor Software growing the team than there is of growing people contributing to MCP. Another of my motivations, why I support this move, is to reduce the burden on MCP.
Roles has been around since I was around Meteor as well, unless something earth shattering comes along in how we think about authorizations then I don't think things are going to change.

tl;dr: I believe that roles inclusion in core makes sense from business perspective if nothing else.

**Reply by harryadel** [reply]
I actually support adding Apple OAuth :+1 . Hey, if hacking/monkey patching lots of stuff to get things working then maybe we're ought to improve the core so others can interface with it more easily and not add more to it. Did I hear someone say [meteorx](https://github.com/monti-apm/meteorx)?

>Another of my motivations, why I support this move, is to reduce the burden on MCP.

EXACTLY!! Brother, if this is the motivation to move it into the core then we're committing a grave mistake. Shifting burdens doesn't mean it'll magically disappear it just makes your problem somebody else problem. And right now Meteor/core has plenty of problems of their own.

**Comment by DanielDornhardt** [comment]
Hi guys!

I'm also kind of wondering what the thought process was to go for `alanning:roles` besides of anything else. Was it because the migration to 3.0 needed some work & so it became an issue who's supporting it along the way?

Or are there other reasons for this?

Because I also think that there are a lot of areas where Meteor could focus on to improve the user experience / quality of Meteor in more central places maybe.

**Reply by DanielDornhardt** [reply]
Or maybe, if they have spare capacity, they could offer a roundtable with MCP now and then to see which packages would need the most support at the given time?

**Reply by jankapunkt** [reply]
I think there was something like this on discord a while ago when they discussed the new roadmap but there were only a few people attending

**Comment by fredmaiaarantes** [comment]
The discussion has been great. @StorytellerCZ already added great points, and I prefer not repeating them. :) But I'd like to highlight some key points.

- The `alanning:roles` package meets important criteria for inclusion in the core: it's widely used, solves a common problem, it's maintainable and enhances the developer experience. These points may be subjective, but it will be the same for any package we consider including.
 
- `alanning:roles` is mature and flexible, covering a wide range of role management needs. Including it in the core would ensure it stays up-to-date with Meteor releases, which could reduce maintenance issues for your projects.
 
- Adding it to the core doesn't prevent the development of alternative solutions. If this package doesn't meet your needs, new packages can be created, or simpler options can be used.

- From an implementation perspective, @StorytellerCZ and @jankapunkt know the package well, which will help the integration process. @denihs will also become familiar with it, as he has previously worked with the accounts package.

- **Remember, our goal is to improve Meteor for all developers.** While Meteor is a full-stack framework, it still lacks some essential features found in community packages. For example, Laravel, which we often compare to Meteor, has built-in authorization. Adonis, Redwood, Blitz, Sails, and Nest do as well.

- Last but not least, we can discuss suggestions on how to improve the package when bringing it to the core. How can we make this package better or simpler?

We still want to add a few more packages to the core. We understand this can lead to different ideas and more discussion. So, for the following packages, we can follow a suggestion made by @nachocodoner:

> One way to make better decisions could be to survey the community about potential packages for the core, allowing people to provide optional explanations for their choices. This would help others understand different perspectives and inspire thoughtful voting. 

I think that’s great, and it can help guide future decisions about what to include in Meteor's core.

**Comment by StorytellerCZ** [comment]
## Meeting notes 10/10/2024
* Current progress
  * [3.1 PR](https://github.com/meteor/meteor/pull/13358)
  * [PR with roles](https://github.com/meteor/meteor/pull/13402)
    * Many thanks to @nachocodoner with the help
* Next steps
  * Removal of deprecated functionality
  * UI Helpers to separate Blaze package
    * React hooks are on the table
  * Docs
    * Need to integrate JSDocs specifics from Meteor into Roles
    * Need create docs page
  * Tests
    * Skipping the discussion for next meeting due to complexity
  * Guide
  * Examples & tutorials
  * Integrations
    * Hooks into user creation
    * Publication of the roles for the current
* Q&A
  * None, post here.

**Comment by jamauro** [comment]
Hey y'all, thanks for your efforts in improving this package.

A few thoughts and questions:

1. Will this be an optional package that someone can add to their project with `meteor add roles` or something similar? I assume so but just double checking since "being in core" could mean different things to people. :)
2. I think this would be a great opportunity to streamline the package as much as possible. IMO, it got kind of bloated over time with several methods to the point where the original author has even said he plans on making a significantly slimmed down version of it. I would encourage y'all to simplify and cut freely. If there is an outcry of a missing method for example, it can always be added back.  
3. Will this be a 3.0+ only package? Are you planning to continue using the `*Async` suffix? Would be nice to move away from it if possible. :)

**Reply by StorytellerCZ** [reply]
1. yes
2. Yes, we will be removing deprecated methods. Right now there are only two and one constant though. The rest will be up for review after the initial merge. I'm open to suggestions.
3. Yes, that was originally the plan for Roles v5 no matter what. The question of `*Async` suffix remains to be resolved as it is a nice indicator between async functions and if missing client safe functions. I'm very much open to suggestions how to best remove the suffix while keeping the convenience of knowing which methods are async and not without having to look it up in the docs.
